### PR TITLE
Add cluster CA certificate into kubeconfig if one exists

### PR DESCRIPTION
**Pull request by ripta**
Addresses one of the ideas in #31. Instead of adding YAML templating, this PR takes a different approach:

1. Check each cluster definition (`clusters[*].cluster` in the kubeconfig) to see if either `certificate-authority-data` (binary embedded PEM data) or `certificate-authority` (path to client-side file) are non-empty in the template.
  a. If both certificate fields are empty and `/var/run/secrets/kubernetes.io/serviceaccount/ca.crt` is accessible by Kuberos (which only happens when Kuberos is run in-cluster, and its pod is given a service account), then the contents of ca.crt is used as the value of `certificate-authority-data`.
  b. Otherwise, no behavior changes.
2. If the template already has a `current-context`, it's still used as the default. Otherwise, `current-context` is set to the first non-empty context.

I'd love feedback overall, but two things jumped out in particular:

3. I'm not sure what the best way to add tests for (1) above, since it needs to test filesystem access. 
4. I added `DefaultAPITokenMountPath`, because there doesn't seem to be an easy way to access the definition of it without pulling in the entirety of `k8s.io/kubernetes/plugin/pkg/admission/serviceaccount`.

### CODE REVIEW

File: kuberos.go [hunk 1]
```diff
@@ -358,9 +365,28 @@ func populateUser(cfg *api.Config, p *extractor.OIDCAuthenticationParams) api.Co
 			},
 		},
 	}
+
 	for name, cluster := range cfg.Clusters {
+		// If the cluster definition does not come with certificate-authority-data nor
+		// certificate-authority, then check if kuberos has access to the cluster's CA
+		// certificate and include it when possible. Assume all errors are non-fatal.
+		if len(cluster.CertificateAuthorityData) == 0 && len(cluster.CertificateAuthority) == 0 {
```
Comment (negz) [inline kuberos.go]: `cluster.CertificateAuthority` is a string, so it would be more idiomatic to test that it's the empty string (i.e. `cluster.CertificateAuthority == ""`.

File: kuberos.go [hunk 2]
```diff
@@ -358,9 +365,28 @@ func populateUser(cfg *api.Config, p *extractor.OIDCAuthenticationParams) api.Co
...
+			caPath := filepath.Join(DefaultAPITokenMountPath, v1.ServiceAccountRootCAKey)
+			if caFile, err := os.Open(caPath); err == nil {
+				if caCert, err := ioutil.ReadAll(caFile); err == nil {
+					cluster.CertificateAuthorityData = caCert
+				}
+			}
+		}
```
Comment (negz) [inline kuberos.go]: https://github.com/spf13/afero

I like to use afero to make this kind of thing testable. It's pretty much a drop in replacement for the stdlib's filesystem functions, with an in-memory backend that's great for testing.

File: kuberos.go [hunk 3]
```diff
@@ -358,9 +365,28 @@ func populateUser(cfg *api.Config, p *extractor.OIDCAuthenticationParams) api.Co
...
 		c.Clusters[name] = cluster
-		c.Contexts[name] = &api.Context{Cluster: name, AuthInfo: p.Username}
+		c.Contexts[name] = &api.Context{
+			Cluster:  name,
+			AuthInfo: p.Username,
+		}
+		// Use the first context as the current context
+		if c.CurrentContext == "" {
+			c.CurrentContext = name
+		}
```
Comment (negz) [inline kuberos.go]: I'm not sure about this functionality. I personally like to avoid setting a current context, because I typically have many clusters and always want to type `kubectl --context=blah` to make sure I never accidentally operate on my current context when I actually meant to operate on another one. (i.e. I want to make it hard to set my current context to a production cluster one day, then assume it's set to the staging cluster the next day and delete a whole namespace). That said I don't use Kuberos in production at my current job, and I think I'm a little weird in this regard, so perhaps this behaviour is what most users would prefer?

**Review by negz — CHANGES_REQUESTED** [review]
I like this approach a lot! I've added some comments - switching to `afero` to make it testable is the only one I feel strongly about.

File: kuberos.go (hunk 1 above)
Comment (ripta) [reply kuberos.go]: Good call! I'll make the change.

File: kuberos.go (hunk 2 above)
Comment (ripta) [reply kuberos.go]: I attempted to use a similar library (`github.com/blang/vfs`) for a work project, and ended up running into a lot of tiny little differences compared to stdlib. 

I'll check out afero, though!

File: kuberos.go (hunk 3 above)
Comment (ripta) [reply kuberos.go]: Yeah, I considered whether or not it's worth it.

We (work) recommend keeping configs for separate clusters in separate configs and setting the `KUBECONFIG` env-var to switch instead. It's obviously a different workflow, so having a default context is useful.

I'll say that this behavior does seem a little magical, though. So maybe I should just remove it, haha.

**Comment by ripta** [comment]
@negz - I added a couple of test cases; used afero, and I really like it so far. Please take a look and LMK what you think!

**Review by negz — APPROVED** [review]
Looks great, thank you!
